### PR TITLE
sql: delete local implementations of planNodes

### LINKED ISSUE #33173: SCRUB index check hangs when run concurrently with TPCC
The new SCRUB roachtests `scrub/{all-checks,index-only}/tpcc-1000`, which run a series of SCRUB checks on a cluster running TPCC at the same time, have been timing out because the SCRUB query hangs. See #33151, #33149.

The TPCC queries themselves run successfully during this test, and the cluster is able to execute other queries when I ssh into one of the nodes. SCRUB also runs fine when run on the cluster with TPCC data with no other queries running, so it seems like the deadlock occurs when there's contention between TPCC queries and the SCRUB queries that need to do a full table scan. (I also tried running the roachtest with `AS OF SYSTEM TIME '-5s'` in the SCRUB query to reduce contention, which was successful. See #33152.)

Goroutine dump: [goroutines.zip](https://github.com/cockroachdb/cockroach/files/2681881/goroutines.zip)

### PR DESCRIPTION
**Pull request by jordanlewis**
This PR deletes the remaining users of the planNode execution engine and deletes the duplicate implementations for those planNodes that have DistSQL equivalents.

Closes #33173.

### CODE REVIEW

**Review by RaduBerinde — COMMENTED** [review]
Wow..  This is fantastic!! I dreamt of this day.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-:-LXQqEAX32YElXTmyZMS:b-74n900)* status: :shipit: complete! 0 of 0 LGTMs obtained

File: pkg/sql/scan.go [hunk 1]
```diff
@@ -193,34 +191,15 @@ func (n *scanNode) IndexedVarNodeFormatter(idx int) tree.NodeFormatter {
...
 type scanRun struct {
-	// Contains values for the current row. There is a 1-1 correspondence
-	// between resultColumns and values in row.
-	row tree.Datums
-
-	// the index of the current row.
-	rowIndex int64
-
-	scanInitialized  bool
 	isSecondaryIndex bool
```
Comment (RaduBerinde) [inline pkg/sql/scan.go]: [nit] Seems like these should be moved to scanNode

**Review by RaduBerinde — APPROVED** [review]

File: pkg/sql/opt_decompose_test.go [hunk 2]
```diff
@@ -72,7 +72,6 @@ func makeSelectNode(t *testing.T, p *planner) *renderNode {
 	numColumns := len(sel.sourceInfo[0].SourceColumns)
 	sel.ivarHelper = tree.MakeIndexedVarHelper(sel, numColumns)
 	p.extendedEvalCtx.IVarContainer = sel
-	sel.run.curSourceRow = make(tree.Datums, numColumns)
```
[the remark this answers is not in the document]
Comment (RaduBerinde) [reply pkg/sql/opt_decompose_test.go]: Ah, sorry, now that I look at it again it was a stupid question :)

**Review by knz — COMMENTED** [review]
Excellent work. It must have been exhilarating!

I have a few minor comments, then this will be good to go.

> Reviewed 17 of 17 files at r1, 6 of 6 files at r2, 37 of 37 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-:-LXYVOA51ewpLPrzy0aO:b-9k4zp6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @RaduBerinde)

___
*[pkg/sql/check.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYRlYEDia2zPeVcVcT:-LXYRlYFDujC-rFzCq-2:b-3tlfhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccef86db95ffa2a2f8321790c687f4ffd3c6bd29/pkg/sql/check.go#L56)):*
> ```Go
> 	}
> 	defer plan.Close(ctx)
> 	rows, err := p.runWithDistSQL(ctx, plan)
> ```

two problems here (2nd commit):
- missing check on `err`
- you must close `rows` even when `rows.Len()` is 0.

So, please:

```
rows, err :=
if err != nil {
   return err
}
defer rows.Close()
if rows.Len() == 0 {
    return nil
}
// ... common case (> 0) here ...
```
___
*[pkg/sql/check.go, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYS98Q0RKXB66jkMPF:-LXYS98RDaYPBieK1gOd:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccef86db95ffa2a2f8321790c687f4ffd3c6bd29/pkg/sql/check.go#L131)):*
> ```Go
> 	defer plan.Close(ctx)
> 
> 	rows, err := p.runWithDistSQL(ctx, plan)
> ```

ditto
___
*[pkg/sql/check.go, line 57 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYTHkg04AV3dUGLjcO:-LXYTHkg04AV3dUGLjcP:b-i9c0pn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8b4534f1ae738addd8e4810dc83273069c28e9e/pkg/sql/check.go#L57)):*
> ```Go
> 	defer plan.Close(ctx)
> 	rows, err := p.runWithDistSQL(ctx, plan)
> 	if err != nil {
> ```

I see you're changing this in the 3rd commit but that's still not good enough. Make the right thing happen in the 2nd commit already.
___
*[pkg/sql/planner.go, line 540 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYSKlc1fFPyhyqBHAC:-LXYSKlc1fFPyhyqBHAD:b-3i1o85) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccef86db95ffa2a2f8321790c687f4ffd3c6bd29/pkg/sql/planner.go#L540)):*
> ```Go
> // In general, you should always prefer to use the internal executor if you can.
> // This method is provided for compatibility with clients that build planNode
> // trees by hand and expect to be able to use the old startPlan method.
> ```

The last sentence should not be included in the comment text - after the PR goes in, this sentence will be meaningless.
___
*[pkg/sql/planner.go, line 544 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYSTkJC7ct22wRsPH7:-LXYSTkK2KQLxC6qKS45:b95hpqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccef86db95ffa2a2f8321790c687f4ffd3c6bd29/pkg/sql/planner.go#L544)):*
> ```Go
> 	ctx context.Context, plan planNode,
> ) (*sqlbase.RowContainer, error) {
> 	params := runParams{
> ```

Please interleave explanatory comments in this code.
___
*[pkg/sql/planner.go, line 581 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYSaP94vSDUVgawW4k:-LXYSaPADy9y7-Ysco9P:b-6ttigr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccef86db95ffa2a2f8321790c687f4ffd3c6bd29/pkg/sql/planner.go#L581)):*
> ```Go
> 		return rows, rowResultWriter.Err()
> 	} else if rows.Len() == 0 {
> 		rows.Close(ctx)
> ```

This is not idiomatic go. Let the caller close rows when err is nil.

As it is now you're checking rows.Len() in two places every time. 
___
*[pkg/sql/scrub_constraint.go, line 118 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYSooVB6bhbNY-zVoD:-LXYSooVB6bhbNY-zVoE:b-idp4u9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccef86db95ffa2a2f8321790c687f4ffd3c6bd29/pkg/sql/scrub_constraint.go#L118)):*
> ```Go
> 	rows, err := scrubRunDistSQL(ctx, planCtx, params.p, physPlan, columnTypes)
> 	if err != nil {
> 		rows.Close(ctx)
> ```

if err is not nil,  `rows` should be nil!?
If that's not the case, you have a seriously non-idiomatic go interface in here.
___
*[pkg/sql/logictest/testdata/planner_test/needed_columns, line 75 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYUzqF040_L8mfD8Hg:-LXYUzqF040_L8mfD8Hh:bkrdcyl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8b4534f1ae738addd8e4810dc83273069c28e9e/pkg/sql/logictest/testdata/planner_test/needed_columns#L75)):*
> ```
> render                   ·         ·  (a)  a=CONST
>  │                       render 0  1  ·    ·
>  └── union               ·         ·  (s)  ·
> ```

1) I think EXPLAIN should still enable the user to distinguish between UNION and UNION ALL, as they are semantically very different. I'm OK if you don't like the word "append" but then please find something else.

2) oy what does the `s` not get omitted?

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34383)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @knz, and @RaduBerinde)

___
*[pkg/sql/check.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYRlYEDia2zPeVcVcT:-LYA8DXm8eGDPnvDclDM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccef86db95ffa2a2f8321790c687f4ffd3c6bd29/pkg/sql/check.go#L56)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

two problems here (2nd commit):
- missing check on `err`
- you must close `rows` even when `rows.Len()` is 0.

So, please:

```
rows, err :=
if err != nil {
   return err
}
defer rows.Close()
if rows.Len() == 0 {
    return nil
}
// ... common case (> 0) here ...
```
</blockquote></details>

Done.
___
*[pkg/sql/check.go, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYS98Q0RKXB66jkMPF:-LYA9MFE6-9_R7E4U5-c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccef86db95ffa2a2f8321790c687f4ffd3c6bd29/pkg/sql/check.go#L131)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[pkg/sql/check.go, line 57 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYTHkg04AV3dUGLjcO:-LYA8DqL9CJpTMnkrQri:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8b4534f1ae738addd8e4810dc83273069c28e9e/pkg/sql/check.go#L57)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I see you're changing this in the 3rd commit but that's still not good enough. Make the right thing happen in the 2nd commit already.
</blockquote></details>

Done.
___
*[pkg/sql/planner.go, line 540 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYSKlc1fFPyhyqBHAC:-LYAAHra6mAgPUrBhvMY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccef86db95ffa2a2f8321790c687f4ffd3c6bd29/pkg/sql/planner.go#L540)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The last sentence should not be included in the comment text - after the PR goes in, this sentence will be meaningless.
</blockquote></details>

Done.
___
*[pkg/sql/planner.go, line 544 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYSTkJC7ct22wRsPH7:-LYABS-81Hmq5PsymnFH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccef86db95ffa2a2f8321790c687f4ffd3c6bd29/pkg/sql/planner.go#L544)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please interleave explanatory comments in this code.
</blockquote></details>

Done.
___
*[pkg/sql/planner.go, line 581 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYSaP94vSDUVgawW4k:-LYAALHV048-jIRdbROj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccef86db95ffa2a2f8321790c687f4ffd3c6bd29/pkg/sql/planner.go#L581)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This is not idiomatic go. Let the caller close rows when err is nil.

As it is now you're checking rows.Len() in two places every time. 
</blockquote></details>

Done.
___
*[pkg/sql/scan.go, line 194 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXQoG7s-nTtQ_VLATxj-r3-194:-LYAAwalAs5Srr2cLTvv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8b4534f1ae738addd8e4810dc83273069c28e9e/pkg/sql/scan.go#L194)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] Seems like these should be moved to scanNode
</blockquote></details>

Done.
___
*[pkg/sql/scrub_constraint.go, line 118 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYSooVB6bhbNY-zVoD:-LYABi3eAjRyfzNv5Iny:bfp4atb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccef86db95ffa2a2f8321790c687f4ffd3c6bd29/pkg/sql/scrub_constraint.go#L118)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

if err is not nil,  `rows` should be nil!?
If that's not the case, you have a seriously non-idiomatic go interface in here.
</blockquote></details>

Not the point of my change. I'll let somebody else fix this.
___
*[pkg/sql/logictest/testdata/planner_test/needed_columns, line 75 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYUzqF040_L8mfD8Hg:-LYAG8ri2VLFnZNFfcKi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8b4534f1ae738addd8e4810dc83273069c28e9e/pkg/sql/logictest/testdata/planner_test/needed_columns#L75)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) I think EXPLAIN should still enable the user to distinguish between UNION and UNION ALL, as they are semantically very different. I'm OK if you don't like the word "append" but then please find something else.

2) oy what does the `s` not get omitted?
</blockquote></details>

Done.

**Review by knz — COMMENTED** [review]
:lgtm_strong:  :tada: 
good riddance, and super well done. Thank you so much!

> Reviewed 7 of 46 files at r4, 5 of 6 files at r5, 37 of 37 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-:-LYChQrqEbwJokslYJRi:b-48cff5)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @knz, and @RaduBerinde)

___
*[pkg/sql/scrub_constraint.go, line 118 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34383#-LXYSooVB6bhbNY-zVoD:-LYCh4FP9KI8WyE8AWZY:bo4udss) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccef86db95ffa2a2f8321790c687f4ffd3c6bd29/pkg/sql/scrub_constraint.go#L118)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Not the point of my change. I'll let somebody else fix this.
</blockquote></details>

:+1: but maybe add a TODO

**Pull request merged**